### PR TITLE
A crud field type to output a VUE.JS component

### PR DESCRIPTION
**Pull request by jrpub**
The goal here is to interface "easily" a CRUD field with a VUE.JS component to be able to do client-side stuff. It supports passing any number of parameters (called "props" in vue JS)

Usage:
`$this->crud->addField([
	'type' => 'vue_component',
	'name' => 'my_vue_js_component',
	'parameter1' => ['test1', 'test2', 'test3'],
	'parameter2' => User::find(1)->toArray()
]);`

In the form page with all the fields, the field will be outputed like that:
`<my_vue_js_component :parameter1:'...' :parameter2:'....'></my_vue_js_component>`

Parameters will be automatically converted to json arrays. 

**Note: All field keys except: 'label', 'name', and 'type' are passed as parameters to the vue.js components.**

Then, in your single-file component, use the parameters you passed as props:
`export default {
        props: ['parameter1', 'parameter2']
	(...)
}`

Reminder, your single-file component should be loaded from app.js:
`Vue.component('my_vue_js_component',  require('./components/my_vue_js_component.vue') );`

Then it will be the responsability of your vue.js component to output what is needed (a dedicated field, or other).

### CODE REVIEW

**Comment by tabacitu** [comment]
Hi @jrpub ,

Thanks a lot for the PR. Sorry for taking so long to answer, we've been knee-deep in issues. I do think we can make it easier to include vue components, but I'm not 100% this is the way to go. Mainly because this is not consistent with the way you _usually_ create a new field. 

Normally to create a new field type you create a new file in ```resources/views/vendor/backpack/crud/fields/``` and have all your JS logic there. If we merge this, we would basically be adding a new way to create field types if they use Vue, with a separate file for the component. **Wouldn't it be a more consistent solution to include all Vue logic inside a blade file, like fields are created now?** In my view, you can totally do that right now, with the solution we have.

Thanks again. Cheers!

**Comment by jrpub** [comment]
Hi everyone, hello @tabacitu 

Ok. a quick question, and a summary of the issue for the next contributors (as I will keep my component for my personal needs as it is extremely used in my projects :-)).

>  Wouldn't it be a more consistent solution to include all Vue logic inside a blade file, like fields are created now? 

Could you explain the design architecture you have in mind? I may have missed something, I don't understand how to do this. I see one issue and three design choices.

== Summary of the issue == 

**Issue**
Integrate complex JS code (managed by a JS framework as it should be for all complex JS code) into Backpack forms/fields easily. 

**Options**

_Option 1_
Stick to pure JS/Jquery and put everything in blade files. 

_Option 2_
Let's put your call to your VueJS component into a new CRUD field blade file. 

Customfield1.blade.php will contain 
`<custom-component :parameter=[...]></custom-component>`

Customfield2.blade.php will contain
 `<custom-component2 :parameter=[...]></custom-component2>`

_Option 3_
Make **one** generic custom CRUD field that would call the VueJS component you want, with the parameters you want (option I selected and tried to make above).

Did I miss another potential possibility?

**Comment by tabacitu** [comment]
Hm.... @OwenMelbz what do you think?

**Comment by OwenMelbz** [comment]
:)

I’ve nothing against the overall idea. HOWEVER.

As vue components need to be transpiled, whether it’s because it’s in a self contained .vue file or because it’s just in es6 - you’d need to make the decision that although it would be supported - it wouldn’t actually “work”

As there’s no asset build pipeline for the backend users would have to set this all up theirselves via webpack/gulp what ever.

So yeah - it wouldn’t work out the box.

Secondly need to think of a mechanism to allow the defining of “parsed” params vs strings e.g

:param=“backback.com” would error
:param=“‘backpack.com’” would work
param=“backpack.com” would work

—

So in summary, sure you can make a field type that literally formats some html. However the whole thing won’t work unless the user includes a backend asset pipeline

**Comment by OwenMelbz** [comment]
Also I think we should use the “attributes” key like all the other components, rather than the properties on the overall field. This also means you can exclude all the in_array malarkey
